presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -273,287 +213,224 @@ public boolean isTypeOnlyCoercion(Type source, Type result)
-            TypeSignature firstType = typeSignaturePair.getFirstType();
-            TypeSignature secondType = typeSignaturePair.getSecondType();
-
-            List<TypeSignatureParameter> firstTypeParameters = firstType.getParameters();
-            List<TypeSignatureParameter> secondTypeParameters = secondType.getParameters();
-
-            checkState(firstTypeParameters.size() == 2);
-            checkState(secondTypeParameters.size() == 2);
-
-            long firstPrecision = firstTypeParameters.get(0).getLongLiteral();
-            long secondPrecision = secondTypeParameters.get(0).getLongLiteral();
-            long firstScale = firstTypeParameters.get(1).getLongLiteral();
-            long secondScale = secondTypeParameters.get(1).getLongLiteral();
-            long targetScale = Math.max(firstScale, secondScale);
-            long targetPrecision = Math.max(firstPrecision - firstScale, secondPrecision - secondScale) + targetScale;
-            targetPrecision = Math.min(38, targetPrecision); //we allow potential loss of precision here. Overflow checking is done in operators.
-            return Optional.of(new TypeSignature(
-                    StandardTypes.DECIMAL,
-                    ImmutableList.of(TypeSignatureParameter.of(targetPrecision), TypeSignatureParameter.of(targetScale))));
+        String sourceTypeName = sourceType.getTypeSignature().getBase();
+        if (sourceTypeName.equals(resultTypeName)) {
+            return Optional.of(sourceType);
+        }
+
+        switch (sourceTypeName) {
+            case UnknownType.NAME: {
+                switch (resultTypeName) {
+                    case StandardTypes.BOOLEAN:
+                    case StandardTypes.BIGINT:
+                    case StandardTypes.INTEGER:
+                    case StandardTypes.DOUBLE:
+                    case StandardTypes.VARBINARY:
+                    case StandardTypes.DATE:
+                    case StandardTypes.TIME:
+                    case StandardTypes.TIME_WITH_TIME_ZONE:
+                    case StandardTypes.TIMESTAMP:
+                    case StandardTypes.TIMESTAMP_WITH_TIME_ZONE:
+                    case StandardTypes.HYPER_LOG_LOG:
+                    case StandardTypes.P4_HYPER_LOG_LOG:
+                    case StandardTypes.JSON:
+                    case StandardTypes.INTERVAL_YEAR_TO_MONTH:
+                    case StandardTypes.INTERVAL_DAY_TO_SECOND:
+                    case RegexpType.NAME:
+                    case LikePatternType.NAME:
+                    case JsonPathType.NAME:
+                    case ColorType.NAME:
+                        return Optional.of(getType(new TypeSignature(resultTypeName)));
+                    case StandardTypes.VARCHAR:
+                        return Optional.of(createVarcharType(0));
+                    case StandardTypes.DECIMAL:
+                        return Optional.of(createDecimalType(1, 0));
```

Comment:
It should coerce to `DECIMAL(0,0)`. 
Otherwise we end up with situation that `CAST(NULL as DECIMAL(x,y))` would not coerce to `DECIMAL(1,1)`.

It is problematic as we disallow DECIMAL(0,0) right now.